### PR TITLE
Imageset

### PR DESCRIPTION
**Pull request by ranjithkumar007**
will rebase it once [this PR](https://github.com/symengine/symengine/pull/1291) gets in.

### CODE REVIEW

**Comment by ranjithkumar007** [comment]
ping @nishnik , @srajangarg .

File: symengine/sets.cpp [hunk 1]
```diff
@@ -716,6 +716,79 @@ RCP<const Boolean> ConditionSet::contains(const RCP<const Basic> &o) const
...
+        int c2 = unified_compare(expr_, other.expr_);
+        if (c2 != 0) {
+            return c2;
+        } else {
+            return unified_compare(base_, other.base_);
+        }
+    }
+}
+
+bool ImageSet::is_canonical(const RCP<const Basic> &sym,
```
Comment (srajangarg) [inline symengine/sets.cpp]: Is it canonical if the `base` is the `EmtpySet`?

File: symengine/tests/basic/test_sets.cpp [hunk 2]
```diff
@@ -734,3 +736,48 @@ TEST_CASE("ConditionSet : Basic", "[basic]")
     REQUIRE(is_a<ConditionSet>(*r1));
     REQUIRE(eq(*down_cast<const ConditionSet &>(*r1).get_condition(), *cond3));
 }
+
+TEST_CASE("ImageSet : Basic", "[basic]")
```
Comment (srajangarg) [inline symengine/tests/basic/test_sets.cpp]: can you add another test case, a more complicated one?

File: symengine/sets.cpp (hunk 1 above)
Comment (ranjithkumar007) [inline symengine/sets.cpp]: no.

**Comment by ranjithkumar007** [comment]
ping @srajangarg

File: symengine/sets.h [hunk 3]
```diff
@@ -326,6 +367,23 @@ inline RCP<const Set> interval(const RCP<const Number> &start,
...
+                               const RCP<const Set> &base)
+{
+    if (not is_a<Symbol>(*sym))
+        throw SymEngineException("first arg is expected to be a symbol");
+
+    if (eq(*expr, *sym) or eq(*base, *emptyset()))
+        return base;
+
+    if (is_a_Number(*expr))
+        return finiteset({expr});
```
Comment (nishnik) [inline symengine/sets.h]: The same would apply for `FiniteSet`.

Comment (ranjithkumar007) [inline symengine/sets.h]: oh yeah!

File: symengine/sets.h [hunk 4]
```diff
@@ -326,6 +367,23 @@ inline RCP<const Set> interval(const RCP<const Number> &start,
 }
 
 // ! \return RCP<const Set>
+inline RCP<const Set> imageset(const RCP<const Basic> &sym,
+                               const RCP<const Basic> &expr,
+                               const RCP<const Set> &base)
+{
+    if (not is_a<Symbol>(*sym))
```
Comment (nishnik) [inline symengine/sets.h]: What if we need more that one symbol?
Like: `for x,y in [0,1] : x**2 + y**2`.
IMO this is doble.

Comment (ranjithkumar007) [inline symengine/sets.h]: I think it's better to implement that feature when it is required.

File: symengine/sets.cpp [hunk 5]
```diff
@@ -716,6 +716,80 @@ RCP<const Boolean> ConditionSet::contains(const RCP<const Basic> &o) const
...
+}
+
+RCP<const Boolean> ImageSet::contains(const RCP<const Basic> &a) const
+{
+    throw SymEngineException("Not implemented");
+}
+
+RCP<const Set> ImageSet::set_union(const RCP<const Set> &o) const
+{
+    return SymEngine::set_union({rcp_from_this_cast<const Set>(), o}, false);
```
Comment (nishnik) [inline symengine/sets.cpp]: We can simplify when `sym` and `base` are same.

File: symengine/sets.cpp [hunk 6]
```diff
@@ -716,6 +716,80 @@ RCP<const Boolean> ConditionSet::contains(const RCP<const Basic> &o) const
...
+}
+
+RCP<const Set> ImageSet::set_intersection(const RCP<const Set> &o) const
+{
+    return SymEngine::set_intersection({rcp_from_this_cast<const Set>(), o});
```
Comment (nishnik) [inline symengine/sets.cpp]: We can simplify when `sym` and `base` are same.

File: symengine/sets.cpp [hunk 7]
```diff
@@ -716,6 +716,80 @@ RCP<const Boolean> ConditionSet::contains(const RCP<const Basic> &o) const
...
+}
+
+RCP<const Set> ImageSet::set_complement(const RCP<const Set> &o) const
+{
+    return SymEngine::set_complement(rcp_from_this_cast<const Set>(), o);
```
Comment (nishnik) [inline symengine/sets.cpp]: We can simplify when `sym` and `base` are same.

File: symengine/sets.cpp (hunk 5 above)
Comment (ranjithkumar007) [inline symengine/sets.cpp]: *`o` and `base`?

Comment (nishnik) [inline symengine/sets.cpp]: For that `o` needs to be an imageset.

Comment (ranjithkumar007) [inline symengine/sets.cpp]: i didn't get it. can you please give an example?

File: symengine/sets.h (hunk 3 above)
Comment (nishnik) [inline symengine/sets.h]: We should follow SymPy here:
image

Comment (ranjithkumar007) [inline symengine/sets.h]: cool. This also holds for other type of sets as well.
```
>>> ImageSet(Lambda(x,Interval(1,y)),x)
{[1, y] | x ∊ x}
>>> ImageSet(Lambda(x,Interval(1,10)),x)
{[1, 10]}
```

File: symengine/sets.cpp (hunk 5 above)
Comment (nishnik) [inline symengine/sets.cpp]: Suppose `o` is an imageset:
`x**2 for x in [0, 1]`
and `this` is:
`x**3 for x in [0, 1]`
Then the union should be:
`Union(x**2, x**3) for x in [0, 1]`

Comment (ranjithkumar007) [inline symengine/sets.cpp]: Okay !

File: symengine/sets.h (hunk 4 above)
Comment (nishnik) [inline symengine/sets.h]: I would like considering @srajangarg's opinion.

File: symengine/sets.cpp (hunk 5 above)
Comment (ranjithkumar007) [inline symengine/sets.cpp]:
```
>>> a=ImageSet(Lambda(x,x**2),S.Reals)
>>> b=ImageSet(Lambda(x,x**3),S.Reals)
>>> Union(a,b)
⎧ 2        ⎫   ⎧ 3        ⎫
⎨x  | x ∊ ℝ⎬ ∪ ⎨x  | x ∊ ℝ⎬
⎩          ⎭   ⎩          ⎭
>>> a.union(b)
⎧ 2        ⎫   ⎧ 3        ⎫
⎨x  | x ∊ ℝ⎬ ∪ ⎨x  | x ∊ ℝ⎬
⎩          ⎭   ⎩          ⎭
```
SymPy doesn't evaluate. Should we continue with evaluating it then ?

Comment (ranjithkumar007) [inline symengine/sets.cpp]: ping @nishnik

Comment (nishnik) [inline symengine/sets.cpp]: leave it then

Comment (ranjithkumar007) [inline symengine/sets.cpp]: peace !

File: symengine/sets.h (hunk 3 above)
Comment (ranjithkumar007) [inline symengine/sets.h]: this feature is added.

**Comment by ranjithkumar007** [comment]
ping @nishnik. can you have another look?

**Comment by srajangarg** [comment]
@nishnik @isuruf any final reviews? we should merge this in.

**Comment by nishnik** [comment]
+1 from me.

**Pull request merged**